### PR TITLE
EROPSPT-169 - Send message to queue when statistics update is required

### PR DESCRIPTION
**Pull request by scl-softwire**

### CODE REVIEW

File: src/main/kotlin/uk/gov/dluhc/printapi/aspect/StatisticsUpdateAspect.kt [hunk 1]
```diff
@@ -0,0 +1,71 @@
...
+        saved.sourceReference?.also {
+            statisticsUpdateService.triggerVoterCardStatisticsUpdate(it)
+        }
+    }
+
+    @AfterReturning(
+        pointcut = "execution(* org.springframework.data.repository.CrudRepository.save(..)) && args(anonymousElectorDocument)",
+        returning = "saved"
+    )
+    fun afterSaveTemporaryCertificate(joinPoint: JoinPoint, anonymousElectorDocument: AnonymousElectorDocument, saved: AnonymousElectorDocument) {
```
Comment (kirstenland) [inline src/main/kotlin/uk/gov/dluhc/printapi/aspect/StatisticsUpdateAspect.kt:41]: Is this named correctly?

File: src/main/kotlin/uk/gov/dluhc/printapi/aspect/StatisticsUpdateAspect.kt [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+        statisticsUpdateService.triggerVoterCardStatisticsUpdate(saved.sourceReference)
+    }
+
+    @AfterReturning(
+        pointcut = "execution(* org.springframework.data.repository.CrudRepository.saveAll(..))",
+        returning = "saved"
+    )
+    fun afterSaveAll(joinPoint: JoinPoint, saved: Iterable<Any>) {
+        // Type is erased and AOP is passed an iterable containing any type
```
Comment (kirstenland) [inline src/main/kotlin/uk/gov/dluhc/printapi/aspect/StatisticsUpdateAspect.kt:50]: I'm not sure I understand this - what's AOP? What type is erased?

File: src/test/kotlin/uk/gov/dluhc/printapi/jobs/BatchPrintRequestsJobIntegrationTest.kt [hunk 3]
```diff
@@ -0,0 +1,94 @@
...
+                    )
+                )
+            )
+        )
+        certificateRepository.save(certificate1)
+        certificateRepository.save(certificate2)
+        TestTransaction.flagForCommit()
+        TestTransaction.end()
+
+        // Clear rogue messages from the queue in order to make the test valid
```
Comment (kirstenland) [inline src/test/kotlin/uk/gov/dluhc/printapi/jobs/BatchPrintRequestsJobIntegrationTest.kt:74]: Rogue maybe makes it sound more worrying than it is - presumably these messages come from creating the certificate in the first place? Is there a reason why this is needed here and not elsewhere?

**Review by kirstenland — COMMENTED** [review]
Looks fairly reasonable to me.

File: src/main/kotlin/uk/gov/dluhc/printapi/aspect/StatisticsUpdateAspect.kt (hunk 2 above)
Comment (scl-softwire) [reply src/main/kotlin/uk/gov/dluhc/printapi/aspect/StatisticsUpdateAspect.kt:50]: Hmm this comment came over from VCA where I ~copied~ got inspiration for this code from. AOP is "Aspect Oriented Programming", the comment is saying that while the `afterSave` methods have explicit types in the signatures this one has an `Any` type, so we need to check types explicitly. Honestly I don't think the comment adds anything - happy to remove?

File: src/test/kotlin/uk/gov/dluhc/printapi/jobs/BatchPrintRequestsJobIntegrationTest.kt (hunk 3 above)
Comment (scl-softwire) [reply src/test/kotlin/uk/gov/dluhc/printapi/jobs/BatchPrintRequestsJobIntegrationTest.kt:74]: Yep, will remove "rogue" and add detail. Calling `certificateRepository.save` is what triggers these messages to go out, so it's only needed on tests where we call that explicitly.

File: src/main/kotlin/uk/gov/dluhc/printapi/aspect/StatisticsUpdateAspect.kt (hunk 2 above)
Comment (kirstenland) [reply src/main/kotlin/uk/gov/dluhc/printapi/aspect/StatisticsUpdateAspect.kt]: Is it worth maybe mentioning how we know all objects are of the same type?

**Review by kirstenland — APPROVED** [review]

**Pull request merged**